design/architecture/api.md

```diff
@@ -19,4 +19,132 @@ and rules on how files and folders are structured and named. Taken
+---
+API security 
+---
+
+Adding a security layer to an API is crucial for ensuring the confidentiality,
+integrity, and availability of the data and systems that use the API. APIs are
+often used to connect different systems and applications, and they provide a way
+for external parties to access and interact with data. Because of this, they can
+be an attractive target for cyber attackers looking to gain unauthorized access,
+steal sensitive information, or disrupt service availability. To protect against
+these threats, it is important to implement a robust security strategy for the
+API.
+
+
+### Approach 1: Basic Authentication
+
+Here is an example for setting up with basic user name password for API request
+
+`````
+class SecureView(APIView):
+    """ Setup secure view of the API with basic authentication """
+    authentication_classes = (BasicAuthentication,)
+    permission_classes = (IsAuthenticated,)
+    
+`````
+
+When make API request need to include the correct username and password
+
+````
+base_url = 'http://example.com/api'
+username = 'username'
+password = 'password'
+
+# Encode the credentials
+credentials = base64.b64encode(f"{username}:{password}".encode()).decode()
+
+# Set the 'Authorization' header to the encoded credentials
+headers = {'Authorization': 'Basic ' + credentials}
+
+# Make the GET request
+response = requests.get(base_url + '/secure-view/', headers=headers)
+````
+
+### Approach 2: User Token for API requests
+
+Here is an example for setting up and token class for API requests (python)
+
+````
+class Token():
+    """ Access token for API requests."""
+    
+    # who created the token
+    user_name = models.CharField()
+    
+    # token field
+    # Use generate token function to create random string
+    token = models.CharField()
+    
+    # This identifies the project or dataset which will be using this token.
+    token_associated_app = models.CharField()
+
+    def generate_token(self):
+        """ populates the token with a random string. """
+
+        charset_string = string.ascii_uppercase + string.ascii_lowercase + string.digits
+        random_string = ''.join([
+            random.SystemRandom().choice(charset_string)
+            for _ in range(TOKEN_LENGTH)
+        ])
+        log.info('regenerate_token: updated token for %s', self.client_name)
+        self.token = random_string
+        self.save()
+
+    def project_associated_with_token(self, project_name):
+        """
+        Verify if the token is associated with the provided app
+        """
+        if self.token_associated_app == project_name:
+            return True
+        return False
+        
+````
+
+When make API request need to include the correct token as the header
+
+````
+base_url = 'http://example.com/api'
+header = { 'Authorization': 'Bearer ' + TOKEN }
+````
+
+### Approach 3: OAuth
+
+Use OAuth approach is more complicated compared to previously two. Here are some
```

Comment:
Haven't actually checked if we want to point out grammar/spelling, if we do, then it should be 'compared to the previous two'.  Otherwise it looks really good.

Reply:
Thank you so much @K-Beicher  :) I will update it on my next commit